File: src/code/DataJam/Repositories/Repository.cs
namespace DataJam;

using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

using JetBrains.Annotations;

/// <summary>Provides an implementation of the Repository pattern.</summary>
/// <param name="dataContext">The data context to use.</param>
[PublicAPI]
public class Repository(IDataContext dataContext) : IRepository
{
    /// <inheritdoc cref="IRepository.Context" />
    public IDataContext Context { get; } = dataContext;

    /// <inheritdoc cref="IRepository.Execute" />
    public void Execute(ICommand command)
    {
        command.Execute(Context);
    }

    /// <inheritdoc cref="IRepository.ExecuteAsync" />
    public Task ExecuteAsync(ICommand command, CancellationToken token = default)
    {
        return Task.Run(() => command.Execute(Context), token);
    }

    /// <inheritdoc cref="IRepository.Find{T}(IQuery{T})" />
    public IEnumerable<T> Find<T>(IQuery<T> query)
    {
        return query.Execute(Context);
    }

    /// <inheritdoc cref="IRepository.Find{T}(IScalar{T})" />
    public T Find<T>(IScalar<T> scalar)
    {
        return scalar.Execute(Context);
    }

    /// <inheritdoc cref="IRepository.FindAsync{T}(IQuery{T},CancellationToken)" />
    public async Task<IEnumerable<T>> FindAsync<T>(IQuery<T> query, CancellationToken token = default)
    {
        return await Task.Run(() => query.Execute(Context), token).ConfigureAwait(false);
    }

    /// <inheritdoc cref="IRepository.FindAsync{T}(IScalar{T}, CancellationToken)" />
    public async Task<T> FindAsync<T>(IScalar<T> scalar, CancellationToken token = default)
    {
        return await Task.Run(() => scalar.Execute(Context), token).ConfigureAwait(false);
    }
}

File: python/src/datajam_sqlalchemy/repository.py
"""SQLAlchemy implementation of IRepository."""

from __future__ import annotations

from typing import TypeVar

from datajam import ICommand, IDataContext, IQuery, IRepository, IScalar

T = TypeVar("T")


class Repository(IRepository):
    """SQLAlchemy implementation of IRepository for command and query execution."""

    def __init__(self, context: IDataContext) -> None:
        """Initialize the repository with a data context."""
        self._context = context

    @property
    def context(self) -> IDataContext:
        """Get the underlying data context."""
        return self._context

    async def execute(self, command: ICommand) -> None:
        """Execute a command."""
        await command.execute(self._context)

    async def find(self, query: IQuery[T]) -> list[T]:
        """Execute a query and return results."""
        return await query.execute(self._context)

    async def find_scalar(self, scalar: IScalar[T]) -> T | None:
        """Execute a scalar query and return the result."""
        return await scalar.execute(self._context)
